src/protocols/http/hpr_http_roaming.erl

```
@@ -0,0 +1,506 @@
+
+    VersionBase = #{
+        'ProtocolVersion' => <<"1.1">>,
+        'SenderNSID' => <<"">>,
+        'DedupWindowSize' => DedupWindowSize
+    },
+
+    VersionBase#{
+        'SenderID' => <<"0xC00053">>,
+        'ReceiverID' => hpr_http_roaming_utils:hexstring(NetID),
+        'TransactionID' => TransactionID,
+        'MessageType' => <<"PRStartReq">>,
+        'PHYPayload' => hpr_http_roaming_utils:binary_to_hexstring(Payload),
+        'ULMetaData' => #{
+            RoutingKey => RoutingValue,
+            'DataRate' => hpr_lorawan:datarate_to_index(Region, DataRate),
+            'ULFreq' => Frequency,
+            'RecvTime' => hpr_http_roaming_utils:format_time(GatewayTime),
+            'RFRegion' => Region,
+            'FNSULToken' => Token,
+            'GWCnt' => erlang:length(Uplinks),
+            'GWInfo' => lists:map(fun gw_info/1, Uplinks)
+        }
+    }.
+
+-spec routing_key_and_value(PacketUp :: hpr_packet_up:packet()) -> {atom(), binary()}.
+routing_key_and_value(PacketUp) ->
+    PacketType = hpr_packet_up:type(PacketUp),
+    {RoutingKey, RoutingValue} =
+        case PacketType of
+            {join_req, {_AppEUI, DevEUI}} ->
+                {'DevEUI', encode_deveui(DevEUI)};
+            {uplink, DevAddr} ->
+                {'DevAddr', encode_devaddr(DevAddr)}
+        end,
+    {RoutingKey, RoutingValue}.
+
+%% ------------------------------------------------------------------
+%% Downlink
+%% ------------------------------------------------------------------
+
+-spec handle_message(prstart_ans() | xmitdata_req()) ->
+    ok
+    | {downlink, xmitdata_ans(), downlink(), {dest_url(), flow_type()}}
+    | {join_accept, downlink()}
+    | {error, any()}.
+handle_message(#{<<"MessageType">> := MT} = M) ->
+    case MT of
+        <<"PRStartAns">> ->
+            handle_prstart_ans(M);
+        <<"XmitDataReq">> ->
+            handle_xmitdata_req(M);
+        _Err ->
+            throw({bad_message, M})
+    end.
```

Do we need a catch all if `M` doesn't have a `<<"MessageType">>` key?

---

We want to fail noisily.